src/parser.rs

```diff
@@ -59,168 +98,231 @@ fn set_details<'a>(
     err
 }
 
-pub(crate) fn document(input: &str) -> IResult<&str, KdlDocument, KdlParseError<&str>> {
-    let (input, nodes) = many0(node)(input)?;
-    let (input, trailing) = all_whitespace(input)?;
-    let mut doc = KdlDocument::new();
-    doc.set_leading("");
-    doc.set_trailing(trailing);
-    *doc.nodes_mut() = nodes;
-    Ok((input, doc))
-}
-
-pub(crate) fn node(input: &str) -> IResult<&str, KdlNode, KdlParseError<&str>> {
-    let (input, leading) = all_whitespace(input)?;
-    let start = input;
-    let (input, ty) = opt(context("valid node type annotation", annotation))(input)?;
-    let (input, name) = context("valid node name", identifier)(input)?;
-    let (input, entries) = many0(context("valid node entry", entry))(input)?;
-    let (input, children) = opt(context("valid node children block", children))(input)?;
-    let (input, trailing) = context(
-        "valid node terminator",
-        cut(recognize(preceded(
-            many0(node_space),
-            alt((
-                terminated(recognize(tag(";")), opt(alt((linespace, eof)))),
-                alt((newline, single_line_comment, eof)),
-            )),
-        ))),
-    )(input)
-    .map_err(|e| {
-        set_details(
-            e,
-            start,
-            Some("parsed node"),
-            Some("Nodes can only be terminated by `;` or a valid line ending."),
-        )
-    })?;
-    let mut node = KdlNode::new(name);
-    node.set_leading(leading);
-    node.set_trailing(trailing);
-    node.ty = ty;
-    let ents = node.entries_mut();
-    *ents = entries;
-    if let Some((before, children)) = children {
-        let childs = node.children_mut();
-        *childs = Some(children);
-        node.set_before_children(before);
+pub(crate) fn document<'a: 'b, 'b>(
+    kdl_parser: &'b KdlParser<'a>,
+) -> impl Fn(&'a str) -> IResult<&'a str, KdlDocument, KdlParseError<&'a str>> + 'b {
+    move |input| {
+        let start = input;
+        let (input, nodes) = many0(node(kdl_parser))(input)?;
+        let (input, trailing) = all_whitespace(kdl_parser)(input)?;
+        let mut doc = KdlDocument::new();
+        doc.set_leading("");
+        doc.set_trailing(trailing);
+        *doc.nodes_mut() = nodes;
+        doc.set_span(kdl_parser.span_from_before_and_after(start, trailing));
```

Comment:
all the spans are setup like this -- we just find the input before parsing the item, and the input after parsing the item, and difference is the span we want.